SimTKcommon/Simulation/include/SimTKcommon/internal/StateImpl.h

```diff
@@ -260,36 +353,102 @@ class CacheEntryInfo {
 
     const AbstractValue& getValue() const {assert(m_value); return *m_value;}
 
-    // Returning the value for write access doesn't affect the version; that
-    // happens only if the cache entries gets marked valid explicitly.
-    AbstractValue& updValue() {assert(m_value); return *m_value;}
+    // Merely handing out the cache entry's value with write access does not
+    // trigger invalidation of dependents. (Maybe it should, but currently it
+    // gets done often with no intent to modify, esp. by SBStateDigest.)
+    // So be sure that the cache entry gets invalidated first either by an
+    // explicit prerequisite change notification, or because the depends-on
+    // stage got invalidated.
+    AbstractValue& updValue(const StateImpl& stateImpl) {
+       assert(m_value); 
+       return *m_value; 
```

Comment:
> Doesn't this violate that references can't be null?

I think this is OK. `m_value` is a pointer. In Debug the pointer is being checked to see that it is non-null. Only then does it get dereferenced. It's true that in Release builds you might get a null reference but lots of bad things can happen in Release builds as a result of various assertions and validity checks being left out; that's a concession to performance. Code really only needs to follow *all* the rules in Debug builds.